### ISSUE TITLE
Detect nullable arguments

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/jnigen#99]
Running JNIgen on `android.database.sqlite` produces non-nullable arguments.

```dart
  /// from: static public android.database.sqlite.SQLiteDatabase openOrCreateDatabase(java.io.File file, android.database.sqlite.SQLiteDatabase.CursorFactory factory)
  /// The returned object must be deleted after use, by calling the `delete` method.
  ///
  /// Equivalent to openDatabase(file.getPath(), factory, CREATE_IF_NECESSARY).
  ///@param file This value must never be {@code null}.
  ///@param factory This value may be {@code null}.
  static SQLiteDatabase openOrCreateDatabase(
          jni.JniObject file, SQLiteDatabase_CursorFactory factory0) =>
```

It would be nice if we could somehow detect arguments that are nullable and provide a nullable type in the signature instead.

Inside the function body we would have to do `factory0 ?? SQLiteDatabase_CursorFactory.fromRef(nullptr)` to pass the C `nullptr` to JNI.

Questions to answer:

* Do Java projects have a standardized way of communicating things are nullable/non-nullable?

**Comment by mahesh-hegde** [comment]
> Do Java projects have a standardized way of communicating things are nullable/non-nullable?

Nope. New projects generally use Optional<T>, but old projects have a few annotations like @NonNull or @Nullable. It seems android has its own.

> Inside the function body we would have to do factory0 ?? SQLiteDatabase_CursorFactory.fromRef(nullptr) to pass the C nullptr to JNI.

We can check at call site before taking the reference. `factory0?.reference ?? nullptr`

**Comment by mahesh-hegde** [comment]
> Hm, maybe we can do a best effort and detect some of the annotations and automatically wrap/unwrap Optional.

I think it's better to just let user specify the fully qualified name of annotation used for nullable or non-null contracts.

In case of API, It can be designed so that a callback can introspect any annotation. It's yet another small tree inside AST :grin:

I don't think anyone is using comments as directives, since annotations were introduced fairly early in the language.

**Comment by dcharkes** [comment]
> I don't think anyone is using comments as directives, since annotations were introduced fairly early in the language.

```
///@param factory This value may be {@code null}.
```

Well, at least the android standard library is. 🙊 

> > Hm, maybe we can do a best effort and detect some of the annotations and automatically wrap/unwrap Optional.
> 
> I think it's better to just let user specify the fully qualified name of annotation used for nullable or non-null contracts.

That doesn't scale.
Are you going to do it manually for all of the java and android APIs? 🙉 

> In case of API, It can be designed so that a callback can introspect any annotation. It's yet another small tree inside AST 😁

Let's first try to make it manually work, and then see how we can generalize it into an API.

**Comment by dcharkes** [comment]
sgtm
